### ISSUE TITLE
Host pattern warnings when using templated target host

### ISSUE THREAD

**Issue by tumbl3w33d** [ansible/ansible#23057]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
Patterns

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel ea7bff4a3f) last updated 2017/03/28 22:53:03 (GMT +200)
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
`Linux milkshake 4.9.0-2-amd64 #1 SMP Debian 4.9.13-1 (2017-02-27) x86_64 GNU/Linux`
`Python 2.7.13`

##### SUMMARY
Using a template variable as argument to `hosts` causes warnings. It seems like the attempt to recognize a pattern expression happens earlier than the variable substitution.

##### STEPS TO REPRODUCE
```yaml
- hosts: "{{ the_host }}"
  tasks:
    - debug:
        msg: "If there's something strange, in the neighborhood. Who you gonna call? bcoca!"
```
Call the playbook and pass the host (in my case my local machine as only entry in the inventory) as extra var
``` bash
ansible-playbook main.yml -e the_host=milkshake -i inventory
```

##### EXPECTED RESULTS
The playbook just runs without any warning and the task gets executed on the targeted host.

##### ACTUAL RESULTS
The playbook runs and the host is properly addressed (ignore the connection failure, that's expected and unrelated), however, ansible prints warnings for no apparent reason.
```
$ ansible-playbook main.yml -e the_host=milkshake -i inventory 
 [WARNING]: Could not match supplied host pattern, ignoring: {{

 [WARNING]: Could not match supplied host pattern, ignoring: the_host

 [WARNING]: Could not match supplied host pattern, ignoring: }}


PLAY [milkshake] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
fatal: [milkshake]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Permission denied (publickey,password).\r\n", "unreachable": true}
```

**Comment by Chris-YBS** [comment]
I can confirm that this is also happening to me. However despite the message only the specified host(s) executes the plays!

BTW I am following a pattern that I found on the web to allow the hosts to be passed either as --extra-vars or from a prompt:

```
- name: ETLDeploy
  vars_prompt:
    - name: "hosts"
      prompt: |- 
                You can just press return for the default or enter a comma seperated
                list of machines with name in uppercase and domain in lower case
                (e.g. XXX.src.com or windowshosts)
                
                Which hosts would you like to setup?
      private: no
      default: "XXX.src.com"
  hosts: "{{hosts}}"
```

**Comment by bcoca** [comment]
so issue here seems to be that we run 'getting hosts' x2, once before templating (you get warnings) and again after (when it works!).

**Comment by bcoca** [comment]
currently fixed in #23001, mostly by changing how inventory is called, will look to see if i can get smaller change to do same

**Comment by ssbarnea** [comment]
I think we need to reopen this bug as I found another case where two warnings are raised without any good reason. This is happening with Ansible 2.4.1 too.

I have some code that looks like:
```
- name: OpenStack Provisioner SSH wait
  hosts: openstack_nodes:bmc:!ovb
  tasks:
  ...
- name: generate inventory file
  hosts: localhost
  gather_facts: no
  tags: always
  tasks:
    - name: update inventory file symlink
      file:
        dest: "{{ inventory_dir }}/hosts"
        state: link
        src: hosts-prov
```
The last tasks throws two warnings:
``` 
[WARNING]: Could not match supplied host pattern, ignoring: bmc
[WARNING]: Could not match supplied host pattern, ignoring: ovb
```

Because this ticket is closed, I had to raise this issue as an individual bug at https://github.com/ansible/ansible/issues/36306 -- better to send your votes/comments there if you want to have it fixed.
